File: src/navigation_mcp/services/trace_callers_service.py
from __future__ import annotations

from pathlib import Path
from typing import Any

from pydantic import ValidationError

from navigation_mcp.adapters.internal_tools import (
    BackendToolError,
    InternalTraceCallersAdapter,
)
from navigation_mcp.contracts.code import (
    PublicFramework,
    PublicLanguage,
    TraceCallerRecord,
    TraceCallersClassificationRecord,
    TraceCallersData,
    TraceCallersImplementationInterface,
    TraceCallersImplementationInterfaceChain,
    TraceCallersImplementationReference,
    TraceCallersInput,
    TraceCallersProbableEntryPoint,
    TraceCallersRecursiveClassifications,
    TraceCallersRecursiveCycle,
    TraceCallersRecursiveData,
    TraceCallersRecursiveNode,
    TraceCallersRecursivePathSegment,
    TraceCallersRecursiveSummary,
    TraceCallersRecursiveTruncatedNode,
    TraceCallersRecursiveVia,
    TraceCallersTarget,
)
from navigation_mcp.contracts.common import (
    ErrorCode,
    ErrorItem,
    ResponseEnvelope,
    ResponseStatus,
)
from navigation_mcp.services.shared import (
    DEFAULT_TRACE_CALLERS_MAX_DEPTH,
    MAX_TRACE_CALLERS_MAX_DEPTH,
    build_response_meta,
    infer_language_from_path,
    prune_recursive_trace_payload,
    resolve_backend_language,
    resolve_effective_language,
    resolve_required_scope,
)

TOOL_NAME = "code.trace_callers"


class TraceCallersService:
    def __init__(
        self, *, workspace_root: Path, adapter: InternalTraceCallersAdapter
    ) -> None:
        self.workspace_root = workspace_root.resolve()
        self.adapter = adapter

    def execute(self, request: TraceCallersInput) -> ResponseEnvelope[TraceCallersData]:
        start_scope = resolve_required_scope(self.workspace_root, request.path)
        start_path = start_scope.relative or Path(".")
        backend_language = resolve_backend_language(request.language, request.framework)
        options = {
            "language": backend_language,
            "recursive": request.recursive,
        }
        if request.recursive:
            options["maxDepth"] = self._resolve_max_depth(request.max_depth)

        internal = self.adapter.trace_callers(
            workspace_root=self.workspace_root,
            path=start_path,
            symbol=request.symbol,
            options=options,
        )

        callers = self._normalize_callers(internal.callers)
        errors: list[ErrorItem] = []
        status = ResponseStatus.OK
        recursive_data = (
            self._normalize_recursive_data(internal.recursive_result)
            if request.recursive and internal.recursive_result
            else None
        )
        recursive_payload_truncated = False
        if recursive_data is not None:
            recursive_data, recursive_payload_truncated = prune_recursive_trace_payload(
                recursive_data
            )
            if recursive_payload_truncated:
                status = ResponseStatus.PARTIAL
                errors.append(
                    ErrorItem(
                        code=ErrorCode.RESULT_TRUNCATED,
                        message="Recursive reverse-trace payload exceeded the V1 response safety caps.",
                        retryable=False,
                        suggestion="Retry with a lower max_depth or disable recursive mode for the direct caller set only.",
                        details={
                            "maxDepth": options.get("maxDepth"),
                            "nodeCount": recursive_data.nodeCount,
                            "pathCount": recursive_data.pathCount,
                        },
                    )
                )

        data = TraceCallersData(
            target=TraceCallersTarget(
                path=start_path.as_posix(),
                symbol=request.symbol,
                language=infer_language_from_path(start_path.as_posix()),
            ),
            count=len(callers),
            returnedCount=len(callers),
            items=callers,
            recursive=recursive_data,
        )

        return ResponseEnvelope[TraceCallersData](
            tool=TOOL_NAME,
            status=status,
            summary=self._build_summary(
                symbol=request.symbol,
                path=start_path.as_posix(),
                count=data.count,
                recursive=request.recursive,
            ),
            data=data,
            errors=errors,
            meta=build_response_meta(
                query=request.model_dump(mode="json"),
                resolved_path=start_scope.public_path,
                truncated=recursive_payload_truncated,
                counts={"returnedCount": data.count, "totalMatched": data.count},
                detection={
                    "effectiveLanguage": (
                        resolve_effective_language(
                            request.language, request.framework
                        ).value
                        if resolve_effective_language(
                            request.language, request.framework
                        )
                        is not None
                        else None
                    ),
                    "framework": request.framework.value if request.framework else None,
                },
            ),
        )

    def error_from_validation(
        self, error: ValidationError
    ) -> ResponseEnvelope[TraceCallersData]:
        details = []
        for item in error.errors():
            details.append(
                {
                    "field": ".".join(str(part) for part in item.get("loc", [])),
                    "message": item.get("msg", "Invalid value."),
                    "type": item.get("type", "validation_error"),
                }
            )

        return ResponseEnvelope[TraceCallersData](
            tool=TOOL_NAME,
            status=ResponseStatus.ERROR,
            summary="Request validation failed.",
            data=self._empty_data(),
            errors=[
                ErrorItem(
                    code=ErrorCode.INVALID_INPUT,
                    message="One or more input fields are invalid.",
                    retryable=False,
                    suggestion="Correct the invalid fields and try again.",
                    details={"issues": details},
                )
            ],
            meta=build_response_meta(query={}),
        )

    def error_from_backend(
        self,
        *,
        request: TraceCallersInput,
        error: BackendToolError,
    ) -> ResponseEnvelope[TraceCallersData]:
        return ResponseEnvelope[TraceCallersData](
            tool=TOOL_NAME,
            status=ResponseStatus.ERROR,
            summary="Caller trace failed.",
            data=self._empty_data(path=request.path, symbol=request.symbol),
            errors=[
                ErrorItem(
                    code=error.code,
                    message=error.message,
                    retryable=error.retryable,
                    suggestion=error.suggestion,
                    details=error.details,
                )
            ],
            meta=build_response_meta(query=request.model_dump(mode="json")),
        )

    def error_from_path(
        self,
        *,
        request_payload: dict[str, Any],
        path_value: str,
        exists: bool,
    ) -> ResponseEnvelope[TraceCallersData]:
        if exists:
            code = ErrorCode.PATH_OUTSIDE_WORKSPACE
            message = f"Path '{path_value}' is outside the configured workspace root."
            suggestion = "Use a file path inside the workspace root."
            summary = "Path validation failed."
        else:
            code = ErrorCode.FILE_NOT_FOUND
            message = f"Path '{path_value}' was not found inside the configured workspace root."
            suggestion = "Provide an existing file path inside the workspace root."
            summary = "Path not found."

        return ResponseEnvelope[TraceCallersData](
            tool=TOOL_NAME,
            status=ResponseStatus.ERROR,
            summary=summary,
            data=self._empty_data(
                path=str(request_payload.get("path") or ""),
                symbol=str(request_payload.get("symbol") or ""),
            ),
            errors=[
                ErrorItem(
                    code=code,
                    message=message,
                    retryable=False,
                    suggestion=suggestion,
                    details={"path": path_value},
                )
            ],
            meta=build_response_meta(query=request_payload),
        )

    def _resolve_max_depth(self, max_depth: int | None) -> int:
        resolved = max_depth or DEFAULT_TRACE_CALLERS_MAX_DEPTH
        return min(resolved, MAX_TRACE_CALLERS_MAX_DEPTH)

    def _normalize_callers(
        self, callers: list[dict[str, Any]]
    ) -> list[TraceCallerRecord]:
        unique: dict[tuple[Any, ...], TraceCallerRecord] = {}
        for item in callers:
            path = str(item.get("file", "")).strip()
            line = int(item.get("line", 1))
            column = item.get("column")
            relation = str(item.get("relation", "unknown")).strip() or "unknown"
            caller = str(item.get("caller", "<unknown>")).strip() or "<unknown>"
            caller_symbol = item.get("traverse_symbol")
            caller_symbol_text = (
                str(caller_symbol).strip() if caller_symbol is not None else None
            )
            key = (path, line, int(column or 0), relation, caller, caller_symbol_text)
            unique[key] = TraceCallerRecord(
                path=path,
                line=line,
                column=int(column) if column is not None else None,
                caller=caller,
                callerSymbol=caller_symbol_text,
                relation=relation,
                language=infer_language_from_path(path),
                snippet=self._normalize_optional_text(item.get("snippet")),
                receiverType=self._normalize_optional_text(item.get("receiverType")),
            )

        return sorted(
            unique.values(),
            key=lambda item: (
                item.path,
                item.line,
                item.column or 0,
                item.callerSymbol or "",
                item.relation,
            ),
        )

    def _normalize_recursive_data(
        self, recursive_result: dict[str, Any]
    ) -> TraceCallersRecursiveData:
        root = recursive_result.get("root", {})
        classifications = recursive_result.get("classifications", {})

        return TraceCallersRecursiveData(
            enabled=bool(recursive_result.get("enabled", True)),
            root=self._normalize_recursive_node(root),
            maxDepth=int(recursive_result.get("maxDepth", 1)),
            maxDepthObserved=int(recursive_result.get("maxDepthObserved", 0)),
            nodeCount=int(recursive_result.get("nodeCount", 0)),
            edgeCount=int(recursive_result.get("edgeCount", 0)),
            pathCount=int(recursive_result.get("pathCount", 0)),
            nodes=[
                self._normalize_recursive_node(node)
                for node in recursive_result.get("nodes", [])
                if isinstance(node, dict)
            ],
            adjacency={
                str(key): [str(child) for child in value]
                for key, value in recursive_result.get("adjacency", {}).items()
                if isinstance(value, list)
            },
            paths=[
                [
                    self._normalize_recursive_path_segment(segment)
                    for segment in path
                    if isinstance(segment, dict)
                ]
                for path in recursive_result.get("paths", [])
                if isinstance(path, list)
            ],
            cycles=[
                TraceCallersRecursiveCycle(
                    fromKey=str(item.get("from", "")),
                    toKey=str(item.get("to", "")),
                    path=[str(part) for part in item.get("path", [])],
                )
                for item in recursive_result.get("cycles", [])
                if isinstance(item, dict)
            ],
            truncated=[
                TraceCallersRecursiveTruncatedNode(
                    key=str(item.get("node", "")),
                    path=str(item.get("file", "")).strip(),
                    symbol=str(item.get("symbol", "")).strip(),
                    depth=int(item.get("depth", 0)),
                )
                for item in recursive_result.get("truncated", [])
                if isinstance(item, dict)
            ],
            probableEntryPoints=[
                self._normalize_probable_entry_point(item)
                for item in recursive_result.get("probableEntryPoints", [])
                if isinstance(item, dict)
            ],
            classifications=TraceCallersRecursiveClassifications(
                summary=TraceCallersRecursiveSummary.model_validate(
                    classifications.get("summary", {})
                ),
                directCallers=[
                    self._normalize_classification_record(item)
                    for item in classifications.get("directCallers", [])
                    if isinstance(item, dict)
                ],
                indirectCallers=[
                    self._normalize_classification_record(item)
                    for item in classifications.get("indirectCallers", [])
                    if isinstance(item, dict)
                ],
                probablePublicEntryPoints=[
                    self._normalize_probable_entry_point(item)
                    for item in classifications.get("probablePublicEntryPoints", [])
                    if isinstance(item, dict)
                ],
                implementationInterfaceChain=[
                    self._normalize_interface_chain(item)
                    for item in classifications.get("implementationInterfaceChain", [])
                    if isinstance(item, dict)
                ],
            ),
        )

    def _normalize_recursive_node(
        self, node: dict[str, Any]
    ) -> TraceCallersRecursiveNode:
        via = node.get("via")
        return TraceCallersRecursiveNode(
            key=str(node.get("key", "")),
            path=str(node.get("file", "")).strip(),
            symbol=str(node.get("symbol", "")).strip(),
            depth=int(node.get("depth", 0)),
            via=TraceCallersRecursiveVia(
                relation=self._normalize_optional_text(via.get("relation")),
                line=int(via.get("line")) if via.get("line") is not None else None,
                column=int(via.get("column"))
                if via.get("column") is not None
                else None,
                snippet=self._normalize_optional_text(via.get("snippet")),
            )
            if isinstance(via, dict)
            else None,
        )

    def _normalize_recursive_path_segment(
        self, segment: dict[str, Any]
    ) -> TraceCallersRecursivePathSegment:
        return TraceCallersRecursivePathSegment(
            path=str(segment.get("file", "")).strip(),
            symbol=str(segment.get("symbol", "")).strip(),
            depth=int(segment.get("depth", 0)),
        )

    def _normalize_probable_entry_point(
        self, item: dict[str, Any]
    ) -> TraceCallersProbableEntryPoint:
        return TraceCallersProbableEntryPoint(
            key=self._normalize_optional_text(item.get("key")),
            path=str(item.get("file", "")).strip(),
            symbol=str(item.get("symbol", "")).strip(),
            depth=int(item.get("depth")) if item.get("depth") is not None else None,
            reasons=[str(reason) for reason in item.get("reasons", [])],
            probable=(
                bool(item.get("probable")) if item.get("probable") is not None else None
            ),
            pathFromTarget=[
                self._normalize_recursive_path_segment(segment)
                for segment in item.get("pathFromTarget", [])
                if isinstance(segment, dict)
            ],
        )

    def _normalize_classification_record(
        self, item: dict[str, Any]
    ) -> TraceCallersClassificationRecord:
        path = str(item.get("file", "")).strip()
        calls = item.get("calls", {})
        return TraceCallersClassificationRecord(
            path=path,
            symbol=str(item.get("symbol", "")).strip(),
            caller=str(item.get("caller", "<unknown>")).strip() or "<unknown>",
            depth=int(item.get("depth", 0)),
            line=int(item.get("line", 1)),
            column=int(item.get("column")) if item.get("column") is not None else None,
            relation=str(item.get("relation", "unknown")).strip() or "unknown",
            language=infer_language_from_path(path),
            receiverType=self._normalize_optional_text(item.get("receiverType")),
            snippet=self._normalize_optional_text(item.get("snippet")),
            calls={
                "path": str(calls.get("file", "")).strip(),
                "symbol": str(calls.get("symbol", "")).strip(),
            },
            pathFromTarget=[
                self._normalize_recursive_path_segment(segment)
                for segment in item.get("pathFromTarget", [])
                if isinstance(segment, dict)
            ],
        )

    def _normalize_interface_chain(
        self, item: dict[str, Any]
    ) -> TraceCallersImplementationInterfaceChain:
        interface = item.get("interface")
        implementation = item.get("implementation")
        implementations = item.get("implementations", [])
        callers = item.get("callers", [])
        return TraceCallersImplementationInterfaceChain(
            kind=str(item.get("kind", "unknown")).strip() or "unknown",
            probable=(
                bool(item.get("probable")) if item.get("probable") is not None else None
            ),
            interface=TraceCallersImplementationInterface(
                name=self._normalize_optional_text(interface.get("name")),
                path=self._normalize_optional_text(interface.get("file")),
                symbol=self._normalize_optional_text(interface.get("symbol")),
            )
            if isinstance(interface, dict)
            else None,
            implementation=TraceCallersImplementationReference(
                path=str(implementation.get("file", "")).strip(),
                symbol=self._normalize_optional_text(implementation.get("symbol")),
            )
            if isinstance(implementation, dict)
            else None,
            implementations=[
                TraceCallersImplementationReference(
                    path=str(entry.get("file", "")).strip(),
                    symbol=self._normalize_optional_text(entry.get("symbol")),
                )
                for entry in implementations
                if isinstance(entry, dict)
            ],
            callers=[
                self._normalize_classification_record(caller)
                for caller in callers
                if isinstance(caller, dict)
            ],
        )

    def _normalize_optional_text(self, value: Any) -> str | None:
        if value is None:
            return None
        normalized = str(value).strip()
        return normalized or None

    def _build_summary(
        self, *, symbol: str, path: str, count: int, recursive: bool
    ) -> str:
        if count == 0:
            return f"Trace completed for incoming callers of '{symbol}' from '{path}' with no callers found."
        if recursive:
            if count == 1:
                return f"Found 1 incoming caller for '{symbol}' from '{path}' with recursive reverse trace."
            return f"Found {count} incoming callers for '{symbol}' from '{path}' with recursive reverse trace."
        if count == 1:
            return f"Found 1 incoming caller for '{symbol}' from '{path}'."
        return f"Found {count} incoming callers for '{symbol}' from '{path}'."

    def _empty_data(self, *, path: str = "", symbol: str = "") -> TraceCallersData:
        return TraceCallersData(
            target=TraceCallersTarget(
                path=path,
                symbol=symbol,
                language=infer_language_from_path(path),
            ),
            count=0,
            returnedCount=0,
            items=[],
            recursive=None,
        )

File: packages/mcp-server/src/services/traceCallersService.ts
import {
  type PublicFramework,
  type PublicLanguage,
  type TraceCallersData,
  type TraceCallersInput,
  normalizeTraceCallersInput,
  type ValidationIssue,
} from "../contracts/public/code.js";
import {
  createResponseMeta,
  type ResponseEnvelope,
} from "../contracts/public/common.js";
import {
  nextRequestId,
  type AnalyzerLanguage,
  type TraceCallersEngineResult,
} from "../engine/protocol.js";
import type { EngineClient } from "../engine/rustEngineClient.js";

const TOOL_NAME = "code.trace_callers";
const DEFAULT_MAX_DEPTH = 3;
const MAX_MAX_DEPTH = 8;

export interface TraceCallersService {
  execute(input: TraceCallersInput): Promise<ResponseEnvelope<TraceCallersData>>;
  validateAndExecute(
    payload: Record<string, unknown>,
  ): Promise<ResponseEnvelope<TraceCallersData>>;
}

export function createTraceCallersService(options: {
  workspaceRoot: string;
  engineClient: EngineClient;
}): TraceCallersService {
  return {
    async execute(input) {
      let response;
      try {
        response = await options.engineClient.request<TraceCallersEngineResult>({
          id: nextRequestId(),
          capability: "workspace.trace_callers",
          workspaceRoot: options.workspaceRoot,
          payload: {
            path: input.path,
            symbol: input.symbol,
            analyzerLanguage: resolveAnalyzerLanguage(input.language, input.framework),
            publicLanguageFilter: resolveEffectiveLanguage(input.language, input.framework),
            recursive: input.recursive,
            maxDepth: input.recursive ? resolveMaxDepth(input.max_depth) : null,
          },
        });
      } catch (error) {
        return buildEngineFailureResponse(input, error);
      }

      if (!response.ok) {
        return buildMappedErrorResponse(
          input,
          response.error.code,
          response.error.message,
          response.error.details,
          response.error.retryable,
        );
      }

      return buildSuccessResponse(input, response.result);
    },
    async validateAndExecute(payload) {
      const normalized = normalizeTraceCallersInput(payload);
      if (!normalized.ok) {
        return buildValidationErrorResponse(normalized.issues);
      }
      return this.execute(normalized.value);
    },
  };
}

function buildSuccessResponse(
  input: TraceCallersInput,
  result: TraceCallersEngineResult,
): ResponseEnvelope<TraceCallersData> {
  const effectiveLanguage = resolveEffectiveLanguage(input.language, input.framework);
  const count = result.totalMatched;

  return {
    tool: TOOL_NAME,
    status: result.truncated ? "partial" : "ok",
    summary: buildSummary(input.symbol, input.path, count, input.recursive),
    data: {
      target: {
        path: input.path,
        symbol: input.symbol,
        language: inferLanguageFromPath(input.path),
      },
      count,
      returnedCount: result.items.length,
      items: result.items,
      recursive: result.recursive,
    },
    errors: result.truncated
      ? [
          {
            code: "RESULT_TRUNCATED",
            message: "Recursive reverse-trace payload exceeded the response safety caps.",
            retryable: false,
            suggestion: "Retry with a lower max_depth or disable recursive mode.",
            details: {
              maxDepth: input.recursive ? resolveMaxDepth(input.max_depth) : null,
            },
          },
        ]
      : [],
    meta: createResponseMeta({
      query: { ...input },
      resolvedPath: result.resolvedPath,
      truncated: result.truncated,
      counts: {
        returnedCount: result.items.length,
        totalMatched: count,
      },
      detection: {
        effectiveLanguage,
        framework: input.framework ?? null,
      },
    }),
  };
}

function buildValidationErrorResponse(
  issues: ValidationIssue[],
): ResponseEnvelope<TraceCallersData> {
  return {
    tool: TOOL_NAME,
    status: "error",
    summary: "Request validation failed.",
    data: emptyData(),
    errors: [
      {
        code: "INVALID_INPUT",
        message: "One or more input fields are invalid.",
        retryable: false,
        suggestion: "Correct the invalid fields and try again.",
        details: { issues },
      },
    ],
    meta: createResponseMeta({ query: {} }),
  };
}

function buildMappedErrorResponse(
  input: TraceCallersInput,
  code: string,
  message: string,
  details: Record<string, unknown>,
  retryable: boolean,
): ResponseEnvelope<TraceCallersData> {
  const query = { ...input };

  if (code === "FILE_NOT_FOUND") {
    return {
      tool: TOOL_NAME,
      status: "error",
      summary: "Path not found.",
      data: emptyData(input.path, input.symbol),
      errors: [
        {
          code,
          message,
          retryable,
          suggestion: "Provide an existing file path inside the workspace root.",
          details,
        },
      ],
      meta: createResponseMeta({ query }),
    };
  }

  if (code === "PATH_OUTSIDE_WORKSPACE") {
    return {
      tool: TOOL_NAME,
      status: "error",
      summary: "Path validation failed.",
      data: emptyData(input.path, input.symbol),
      errors: [
        {
          code,
          message,
          retryable,
          suggestion: "Use a file path inside the workspace root.",
          details,
        },
      ],
      meta: createResponseMeta({ query }),
    };
  }

  if (code === "UNSUPPORTED_CAPABILITY") {
    return {
      tool: TOOL_NAME,
      status: "error",
      summary: "Caller trace failed.",
      data: emptyData(input.path, input.symbol),
      errors: [
        {
          code: "BACKEND_EXECUTION_FAILED",
          message,
          retryable,
          suggestion: "Verify the engine supports workspace.trace_callers and retry.",
          details,
        },
      ],
      meta: createResponseMeta({ query }),
    };
  }

  return {
    tool: TOOL_NAME,
    status: "error",
    summary: "Caller trace failed.",
    data: emptyData(input.path, input.symbol),
    errors: [
      {
        code: code === "BACKEND_EXECUTION_FAILED" ? code : "BACKEND_EXECUTION_FAILED",
        message,
        retryable,
        details,
      },
    ],
    meta: createResponseMeta({ query }),
  };
}

function buildEngineFailureResponse(
  input: TraceCallersInput,
  error: unknown,
): ResponseEnvelope<TraceCallersData> {
  return buildMappedErrorResponse(
    input,
    "BACKEND_EXECUTION_FAILED",
    error instanceof Error ? error.message : String(error),
    {},
    false,
  );
}

function emptyData(path = "", symbol = ""): TraceCallersData {
  return {
    target: {
      path,
      symbol,
      language: inferLanguageFromPath(path),
    },
    count: 0,
    returnedCount: 0,
    items: [],
    recursive: null,
  };
}

function buildSummary(
  symbol: string,
  path: string,
  count: number,
  recursive: boolean,
): string {
  if (count === 0) {
    return `Trace completed for incoming callers of '${symbol}' from '${path}' with no callers found.`;
  }
  if (recursive) {
    if (count === 1) {
      return `Found 1 incoming caller for '${symbol}' from '${path}' with recursive reverse trace.`;
    }
    return `Found ${count} incoming callers for '${symbol}' from '${path}' with recursive reverse trace.`;
  }
  if (count === 1) {
    return `Found 1 incoming caller for '${symbol}' from '${path}'.`;
  }
  return `Found ${count} incoming callers for '${symbol}' from '${path}'.`;
}

function resolveMaxDepth(maxDepth: number | null | undefined): number {
  return Math.min(maxDepth ?? DEFAULT_MAX_DEPTH, MAX_MAX_DEPTH);
}

function resolveEffectiveLanguage(
  language: PublicLanguage | null | undefined,
  framework: PublicFramework | null | undefined,
): PublicLanguage | null {
  if (language) {
    return language;
  }
  if (framework === "react-router") {
    return "typescript";
  }
  if (framework === "spring") {
    return "java";
  }
  return null;
}

function resolveAnalyzerLanguage(
  language: PublicLanguage | null | undefined,
  framework: PublicFramework | null | undefined,
): AnalyzerLanguage {
  const effective = resolveEffectiveLanguage(language, framework);

  if (effective === "java") {
    return "java";
  }
  if (effective === "python") {
    return "python";
  }
  if (effective === "rust") {
    return "rust";
  }
  return "typescript";
}

function inferLanguageFromPath(path: string): PublicLanguage | null {
  const normalized = path.toLowerCase();
  if (normalized.endsWith(".ts") || normalized.endsWith(".tsx")) {
    return "typescript";
  }
  if (normalized.endsWith(".js") || normalized.endsWith(".jsx")) {
    return "javascript";
  }
  if (normalized.endsWith(".java")) {
    return "java";
  }
  if (normalized.endsWith(".py")) {
    return "python";
  }
  if (normalized.endsWith(".rs")) {
    return "rust";
  }
  return null;
}
